KEP-5284: Limit impersonate permissions (#5285)
kubernetes/enhancements: keps/sig-auth/5284-constrained-impersonation/README.md

* Provide clear example on what permission the impersonator and user should have for the
  impersonator to impersonate user's certain verbs on certain resources.
* This is an opt-in feature and existing impersonation flows keep working as-is.
* Client side semantics of impersonation is unchanged, the existing impersonation headers
are re-used and no new headers are added.

### Non-Goals
* Adding impersonator's info to user.Info, so authorization webhooks can know both info
of impersonated user and the requester.

## Proposal

Introduce a set of verbs with prefix of `impersonate-on:`, e.g. `impersonate-on:create` and
`impersonate-on:get`. The impersonator needs to have these verbs with certain resources to
impersonate. 

Introduce verbs `impersonate:user-info`, `impersonate:serviceaccount` and `impersonate:scheduled-node`:
- `impersonate:user-info` limits the impersonator to impersonate users with 
certain names/groups/userextras. The resources must be `users`/`groups`/`userextras`/`uids`.
The resource names must be user names, group names or values in the user extras accoringly.
- `impersonate:serviceaccount` that limits the impersonator to impersonate the serviceaccount with

> the certain name/namespace. The resources must be `serviceaccounts`.

Reviewer: Does `impersonate:serviceaccount` for some resourceName imply access to the the `system:serviceaccounts:` groups associated with those serviceaccounts as well?
Does this verb behave differently than the current `impersonate` verb when `serviceaccounts` is supplied as the resource?

Author: It is the same as `impersonate` verb with `serviceaccount` as the resource, but it will require additional `impersonate-on:{action}` verb. The reason that not to reuse the original impersonate verb is to make it unchanged for backward compatible.

-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ConstrainedImpersonation
  - Components depending on the feature gate:
    - kube-apiserver

###### Does enabling the feature change any default behavior?

No. Impersonator with existing impersonate permission will still be allowed to impersonate

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

Yes. Set the feature gate to false and restart the kube-apiserver.

###### What happens if we reenable the feature if it was previously rolled back?

No addtional configuration is needed.

> ###### Are there any tests for feature enablement/disablement?

Author: yes, updated. This will be tested in ut and integration.

1. A controller impersonates a node the controller is running on. This happens when 
per-node agents (CNI plugins for instance), want to read pods on a given node instead
of unrestricted pod access.
2. A controller runs as a deputy, receives request from the user, and impersonates
the user to start/stop virtual machine or access vm console managed by kubevirt.

This proposal is to introduce additional permissions for impersonation, so that any 
impersonator can impersonate in a more restricted way.

### Goals
* Define required permission rules for the impersonator to impersonate 
  on certain resource/verb.
* Provide clear example on what permission the impersonator and user should have for the
  impersonator to impersonate user's certain verbs on certain resources.
* This is an opt-in feature and existing impersonation flows keep working as-is.
* Client side semantics of impersonation is unchanged, the existing impersonation headers
are re-used and no new headers are added.

### Non-Goals
* Adding impersonator's info to user.Info, so authorization webhooks can know both info
of impersonated user and the requester.

## Proposal

> 

Reviewer 2: If a cluster had its own authn and splits principals into ServiceAccounts, ProjectRoles (not K8s RBAC) and ProjectUsers: does this mechanism prevent things working for that design?

We should not (almost **must not**) assume that clusters rely on K8s RBAC.

Reviewer 3: > I do not think it prevent things working. This still work if it is not k8s RBAC, the external authorizer will receive subject access review call with the specified verb.

Agreed.  The SAR below is well defined and follows our existing semantics.

##### Integration tests

<!--
Integration tests are contained in https://git.k8s.io/kubernetes/test/integration.
Integration tests allow control of the configuration parameters used to start the binaries under test.
This is different from e2e tests which do not allow configuration of parameters.
Doing this allows testing non-default options and multiple different and potentially conflicting command line options.
For more details, see https://github.com/kubernetes/community/blob/master/contributors/devel/sig-testing/testing-strategy.md

If integration tests are not necessary or useful, explain why.
-->

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, document that tests have been written,
have been executed regularly, and have been stable.
This can be done with:
- permalinks to the GitHub source code
- links to the periodic job (typically https://testgrid.k8s.io/sig-release-master-blocking#integration-master), filtered by the test name
- a search in the Kubernetes bug triage tool (https://storage.googleapis.com/k8s-triage/index.html)
-->

> - SAR check on impersonate user with permission:

Reviewer 2: ```suggestion
- SubjectAccessReview check on impersonating user. For RBAC authz mode, this might look like:

```

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: impersonate-user
rules:
- apiGroups:
  - auhtentications.k8s.io
  resources:
  - users
  resourceNames:
  - bob
  verbs:
  - impersonate:user-info
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: impersonate:vm:console
  namespace: default
rules:
  - apiGroups:
    - ""
    resources:
    - nodes
    verbs:
    - impersonate-on:list
```
  - The impersonator can impersonate bob.
  - The impersonator cannot impersonate alice.
  - The impersonator can impersonate on listing nodes
  - The impersonator cannot impersonate on updating nodes

- SAR check on impersonate scheduled node with permissions. The impersonator has the

> user extra info of `"authentication.kubernetes.io/node-name": "node1"`

Reviewer 2: Can any client provide this? What protects against abuse?

The important examples here are actually:
• a compromised node
• an attacker aiming to elevate privileges by impersonating a node

Author: If an attacker intends to elevate privileges by impersonating a node, it needs to have an user extra `"authentication.kubernetes.io/node-name": "someNode"` in the userInfo, and it needs to have both permissions with verbs "impersonated:scheduled-nodes" and "impersonated:someVerb".

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: impersonate-user
rules:
- apiGroups:

>   - auhtentications.k8s.io

Reviewer 2: (typo)

[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
-->
#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Determine if additional tests are necessary
- Ensure reliability of existing tests

#### GA

- At least one successful adoption of each user story (node agent and deputy).
- All bugs resolved and no new bugs requiring code change since the previous shipped release.

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

On upgrade to a version that enables the feature

> * the previous inpersonator with impersonate permission will still work, but it is highly

Reviewer 2: (typo)

team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: ConstrainedImpersonation
  - Components depending on the feature gate:
    - kube-apiserver

###### Does enabling the feature change any default behavior?

No. Impersonator with existing impersonate permission will still be allowed to impersonate

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

> Yes. Set the FeatureGate to false and restart the kube-apiserver.

Reviewer 2: nit: feature gate (2 words)

with and without the feature, are necessary. At the very least, think about
conversion tests if API types are being modified.

Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

There is not impact on rollout, the impersonator with existing impersonate permission can still perform the action.
When the system rollback, impersonator with `impersonate-on:` and `impersonate:` permission will no longer
be authorized to impersonate. Impersonator will need to have the unscoped impersonate permission.

###### What specific metrics should inform a rollback?

> <!--

Reviewer 2: Authz latency?

What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

<!--
Describe manual testing that was done and the outcomes.
Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

<!--
Even if applying deprecation policies, they may still surprise some users.
-->

No

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

> None

Reviewer 2: This surprises me. I would track how many SubjectAccessReviews have been done, or a counter of summed review latency.

Author: listed the existing metrics for authorization latency.

###### How can an operator determine if the feature is in use by workloads?

> <!--

Reviewer 2: Needs specifying at some point.

Reviewer 3: we may not actually have a metric for this yet.  I could see some value in a metric counting impersonation requests and count of failures being useful.

and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->
No.

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

<!--
Describe them, providing:
  - API call type (e.g. PATCH pods)
  - estimated throughput
  - originating component(s) (e.g. Kubelet, Feature-X-controller)
Focusing mostly on:
  - components listing and/or watching resources they didn't before
  - API calls that may be triggered by changes of some Kubernetes resources
    (e.g. update of object X triggers new updates of object Y)
  - periodic API calls to reconcile state (e.g. periodic fetching state,
    heartbeats, leader election, etc.)
-->

> No.

Reviewer 2: What about the additional SubjectAccessReviews? Could there ever be more of these than before?

Reviewer 3: Yes, it should also be listed here.

Author: ok, updated

###### Will enabling / using this feature result in introducing new API types?

<!--
Describe them, providing:
  - API type
  - Supported number of objects per cluster
  - Supported number of objects per namespace (for namespace-scoped objects)
-->
No.

###### Will enabling / using this feature result in any new calls to the cloud provider?

<!--
Describe them, providing:
  - Which API(s):
  - Estimated increase:
-->
No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

<!--
Describe them, providing:
  - API type(s):
  - Estimated increase in size: (e.g., new annotation of size 32B)
  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
-->
No.

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

<!--
Look at the [existing SLIs/SLOs].

Think about adding additional work or introducing new steps in between
(e.g. need to do X to start a container), etc. Please describe the details.

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

> It will introduce one additional access review check for request with impersonation.

Reviewer 2: One? Could it ever be more than one?

    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

## Implementation History

<!--
Major milestones in the lifecycle of a KEP should be tracked in this section.
Major milestones might include:
- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
- the `Proposal` section being merged, signaling agreement on a proposed design
- the date implementation started
- the first Kubernetes release where an initial version of the KEP was available
- the version of Kubernetes where the KEP graduated to general availability
- when the KEP was retired or superseded
-->

## Drawbacks

<!--

> Why should this KEP _not_ be implemented?

Reviewer 2: List these; it's an important part of the KEP.

Author: added

-->

## Alternatives

### Subject Access Review

The controller can sends a SAR request, and then uses its own permission to perform the action.
The main difference from impersonation is:

> 1. the controller itself needs to have the permission for a certain action, while with impersonataion

Reviewer 2: (typo)

the controller does not need these permissions, but the permissions to `impersonate-on` certain action.
2. The audit log shows that controller performs the action, while with impersonatation audit log
shows controller is impersonating and the target user performs the action.
3. The admission chain is running against the controller, while with impersonation, the admission chain
is running against the target user.

### Setting a special APIGroup suffix instead of special verb

Instead of using a verb with prefix `impersonate-on:`, a special apigroup suffix/prefix can be set for
each resource to be impersonated, e.g.

```yaml
- apiGroups:

>   - apps.imperonsation.k8s.io

Reviewer 2: (typo); also, I don't see a suffix here.

  resources:
  - deployments
  verbs:
  - list
  - watch
```

It is almost the same as the verb based approach in the proposal. However, since existing impersonation flows are
verb based, so making the new flow verb based as well feels more consistent.

> ### Check permission intersaction of impersonator and target user

Reviewer 2: (typo)

This is an approach to check intersected permission of the impersonator and the target user, and
only allow the action if both have the correct permission. Comparing to the proposed approach:
this approach requires the impersonator to have the permission who is not desired to have, while
in the proposed apporch, the impersonator's permission is clearer that it can only perform
the action when impersonating.

> ### Expand RBAC/SAR

Reviewer 2: If the cluster does not use RBAC, is this alternative still feasible?

Introduce additional API to define more fine grained access control rule, and ref the rule in SAR.
One example is

```yaml
{
  "apiVersion": "authorization.k8s.io/v1beta1",
  "kind": "SubjectAccessReview",
  "spec": {
    "resourceAttributes": {
      "namespace": "default",
      "verb": "get",
      "group": "example.org",
      "resource": "something"
    },
    "accessRule": {
      "kind": "ClusterAccessRule",
      "group version": "example.org/v1",
      "name": "elevation-of-privilege"
    },
    "user": "lmktfy",

>     "group": []

Reviewer 2: Try not to use real usernames in examples (people sometimes copy them verbatim!)

- kind: ServiceAccount
  name: default
  namespace: default
```

#### Story 2

As a controller, I am working as a deputy, receiving any user's request to access virtual machine console.

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: impersonate-user
rules:
- apiGroups:
  - authentications.k8s.io
  resources:
  - users
  verbs:
  - impersonate:user-info
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: impersonate-user
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: impersonate-user
subjects:
- kind: ServiceAccount
  name: deputy
  namespace: deputy-ns
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: impersonate:vm:console
  namespace: default
rules:
  - apiGroups:
    - subresources.kubevirt.io
    resources:
    - virtualmachines/console
    verbs:
    - impersonate-on:get
---
apiVersion: rbac.authorization.k8s.io/v1
kind: RoleBinding
metadata:
  name: impersonate-user
  namespace: default
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: Role
  name: impersonate:vm:console
subjects:
- kind: ServiceAccount
  name: deputy
  namespace: deputy-ns
```

> ### Risks and Mitigations

Author: Add a section on this.

> ### Risks and Mitigations

Reviewer 4: I think there's potential for more than 2 extra checks:

If impersonation headers are for non-serviceaccount/non-node username with N username+groups+uid+extra attributes, there are now up to 2N + 1 authz checks the filter has to do (N+1 more than before)?
* impersonate:user-info (N checks) && impersonate-on:...
* impersonate (legacy) (N checks)

If impersonation headers are for a service account, there are now up to 3 authz checks the filter has to do (2 more than before)?
* impersonate:serviceaccount && impersonate-on:...
* impersonate (legacy)

If impersonation headers are for a node, there are now up to 4 authz checks the filter has to do (3 more than before)?
* (impersonate:scheduled-node || impersonate:node) && impersonate-on:...
* impersonate (legacy) - N checks for the various facets

Are we just documenting the extra load, or do we have benchmarks / targets that we consider acceptable for this that would inform whether we need to optimize authz paths (rbac indexing or batch authz or something) as part of graduating this?

Currently, all existing working impersonation configurations are using the legacy system. If this KEP switches to do constrained authz checks first, it immediately doubles the impersonation authorization cost for all consumers. As adoption of constrained impersonation increases, I'm not sure how we'll decide which check to place first (constrained or unconstrained) to be the most efficient.

> ### Risks and Mitigations

Reviewer 5: FWIW a 10 second cache like we have for tokens seems like it would be helpful here?

Something like `cache-key(incoming user, impersonation headers) -> impersonated user info`?

of `system:node:` on the request.
2. The user info of the impersonator has an extra with key `authentication.kubernetes.io/node-name`, and the value
should be the same as the value in the request header of `Impersonate-User` after removing prefix `system:node:`.
It indicates that the impersonator is running on the same node it is impersonating.

If the two conditions are both met, the following subjectaccessreview will be sent to the authorizer:
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentications.k8s.io
    resource: nodes
    verb: impersonate:scheduled-node
  user: impersonator
```
If one of the conditions is not met, verb `impersonate:scheduled-node` will not be checked and verb
`impersonate:user-info` will be checked instead. For instance, if condition 2 is not met, the following
subjectaccessreview will be sent.
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentications.k8s.io
    resource: users
    name: system:node:node1
    verb: impersonate:user-info
  user: impersonator
```

### Auditing

> An audit event is recorded already for any allowed impersonation request. To record the reason why the impersonation

Reviewer 5: ```suggestion
Audit events already contain the `impersonatedUser` field to denote if impersonation was used. To record the reason why the impersonation
```

is allowed, an annotation `allowed-impersonation-verbs` will be added. The value will be the list of impersonation

> related verbs checked. For example, `allowed-impersonation-verbs: impersonate:scheduled-node,impersonate-on:list`

Reviewer 5: I don't think we need the `impersonate-on:list` bit because that can be inferred from the request itself right?

are missing a bunch of machinery and tooling and can't do that now.
-->

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

<!--
Even if applying deprecation policies, they may still surprise some users.
-->

No

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

There are existing metrics to record authz latency and request number:
- authorization latency
- authorization success
- webhook authorizer match condition latency
- webhook authorizer match condition success

###### How can an operator determine if the feature is in use by workloads?

<!--
Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
checking if there are objects with field X set) may be a last resort. Avoid
logs or events for this purpose.
-->

It should be user to set the permissions for the controllers. The operator should
not set these permissions for the workload.

> ###### How can someone using this feature know that it is working for their instance?

Reviewer 6: Your doc mentions the use of events. Are there events here or not?

Author: it is audit events which will be finally in the audit log by default.

and creating new ones, as well as about cluster-level services (e.g. DNS):
  - [Dependency name]
    - Usage description:
      - Impact of its outage on the feature:
      - Impact of its degraded performance or high-error rates on the feature:
-->
No.

### Scalability

<!--
For alpha, this section is encouraged: reviewers should consider these questions
and attempt to answer them.

For beta, this section is required: reviewers must answer these questions.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### Will enabling / using this feature result in any new API calls?

<!--
Describe them, providing:
  - API call type (e.g. PATCH pods)
  - estimated throughput
  - originating component(s) (e.g. Kubelet, Feature-X-controller)
Focusing mostly on:
  - components listing and/or watching resources they didn't before
  - API calls that may be triggered by changes of some Kubernetes resources
    (e.g. update of object X triggers new updates of object Y)
  - periodic API calls to reconcile state (e.g. periodic fetching state,
    heartbeats, leader election, etc.)
-->
No.

> ###### Will enabling / using this feature result in introducing new API types?

Reviewer 6: Are you introducing a new verb?

<!--
Describe them, providing:
  - API type
  - Supported number of objects per cluster
  - Supported number of objects per namespace (for namespace-scoped objects)
-->
No.

###### Will enabling / using this feature result in any new calls to the cloud provider?

<!--
Describe them, providing:
  - Which API(s):
  - Estimated increase:
-->
No.

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

<!--
Describe them, providing:
  - API type(s):
  - Estimated increase in size: (e.g., new annotation of size 32B)
  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
-->

> No.

Reviewer 6: Are you not introducing a new verb?

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

<!--
Look at the [existing SLIs/SLOs].

Think about adding additional work or introducing new steps in between
(e.g. need to do X to start a container), etc. Please describe the details.

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

The existing impersonation mechanism will introduce 1 access review for request with impersonation. While upon
feature is enabled:
- When the check is passed, it will introduce 2-3 access review checks for request with impersonation.
  - 2 access review check if the new access rule is passed.
  - 3 access review check if the new access rule is not passed, and the legacy impersonate access rule is passed.
- When the check is disallowed, it will introduce 3 access review checks for request with impersonation.

> ###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

Reviewer 3: N/A doesn't seem like the right answer.  Possibly indicating that requests using this require additional authorization calls, but are expected to be a minority of total requests.

-->

###### Can enabling / using this feature result in resource exhaustion of some node resources (PIDs, sockets, inodes, etc.)?

<!--
Focus not just on happy cases, but primarily on more pathological cases
(e.g. probes taking a minute instead of milliseconds, failed pods consuming resources, etc.).
If any of the resources can be exhausted, how this is mitigated with the existing limits
(e.g. pods per node) or new limits added by this KEP?

Are there any tests that were run/should be run to understand performance characteristics better
and validate the declared limits?
-->
No, this feature does not touch nodes.

### Troubleshooting

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.

The Troubleshooting section currently serves the `Playbook` role. We may consider
splitting it into a dedicated `Playbook` document (potentially with some monitoring
details). For now, we leave it here.
-->

###### How does this feature react if the API server and/or etcd is unavailable?
This feature is fully contained within the API server.

> ###### What are other known failure modes?

Author: added one item.

<!--
For each of them, fill in the following information by copying the below template:
  - [Failure mode brief description]
    - Detection: How can it be detected via metrics? Stated another way:
      how can an operator troubleshoot without logging into a master or worker node?
    - Mitigations: What can be done to stop the bleeding, especially for already
      running user workloads?
    - Diagnostics: What are the useful log messages and their required logging
      levels that could help debug the issue?
      Not required until feature graduated to beta.
    - Testing: Are there any tests for failure mode? If not, describe why.
-->

###### What steps should be taken if SLOs are not being met to determine the problem?

> ## Implementation History

Author: added

should be the same as the value in the request header of `Impersonate-User` after removing prefix `system:node:`.
It indicates that the impersonator is running on the same node it is impersonating.

The flow for checking these two verbs will be as following:
1. If condition 1 is not met, verb `impersonate:user-info` will be checked instead.
2. If both conditions are met, the verb `impersonate:scheduled-node` will be checked at first:
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentications.k8s.io
    resource: nodes
    verb: impersonate:scheduled-node
  user: impersonator
```
3. If check in step 2 is not passed, or only condition 1 is met, the verb impersonate:node will be checked:
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentications.k8s.io
    resource: nodes
    name: node1
    verb: impersonate:node
  user: impersonator
```

### Auditing

Audit events already contain the `impersonatedUser` field to denote if impersonation was used.

> To record the reason why the impersonation is allowed, an annotation `allowed-impersonation-verbs` will

Reviewer 3: Still outstanding.  Let's add a field `impersonationConstraint` that holds the value described here.

- `impersonate:node` that limits the impersonator to impersonate the node only. The resource
must be `nodes`, and the resourceName should be the name of the node. The impersonator must have this
verb to impersonate a node.
- `impersonate:scheduled-node` that limits the impersonator to impersonate the node the 
impersonator is running on. The resources must be `nodes`.

For clusters that use RBAC authz mode, two permissions will be required for impersonation. For example:
to express "system:serviceaccount:default:default can impersonate a user named someUser solely to list
and watch pods in the default namespace."
1. The permission to constrained impersonate a certain user. This is a cluster scoped permission.
```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: constrained-impersonate-only-someUser
rules:
- apiGroups:
  - authentications.k8s.io
  resources:
  - users # allowed resources are users/groups/userextras/uids
  resourceNames:
  - someUser 
  verbs:
  - impersonate:user-info
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: impersonate
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole

>   name: impersonate

Reviewer 3: ```suggestion
  name: constrained-impersonate-only-someUser
```

-->
#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Determine if additional tests are necessary
- Ensure reliability of existing tests
- Determine if some mechanism should be introduced to reduce the extra permission checks

#### GA

- At least one successful adoption of each user story (node agent and deputy).
- All bugs resolved and no new bugs requiring code change since the previous shipped release.
- Conformance tests are added.

### Upgrade / Downgrade Strategy

<!--
If applicable, how will the component be upgraded and downgraded? Make sure
this is in the test plan.

Consider the following in developing an upgrade/downgrade strategy for this
enhancement:
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

On upgrade to a version that enables the feature
* the previous impersonator with impersonate permission will still work, but it is highly

> recommanded to use the new permissions with less privilege.

Reviewer 3: ```suggestion
recommended to use the new permissions with less privilege.
```

are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

Yes, this will be covered in the unit tests and integration tests.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

There is not impact on rollout, the impersonator with existing impersonate permission can still perform the action.
When the system rollback, impersonator with `impersonate-on:` and `impersonate:` permission will no longer
be authorized to impersonate. Impersonator will need to have the unscoped impersonate permission.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

Authz latency for impersonation is too long.

> ###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

Reviewer 3: needs answer.  Without any persisted state, there is no lasting history to clear.

<!--
**Note:** This checklist is iterative and should be reviewed and updated every time this enhancement is being considered for a milestone.
-->

[kubernetes.io]: https://kubernetes.io/
[kubernetes/enhancements]: https://git.k8s.io/enhancements
[kubernetes/kubernetes]: https://git.k8s.io/kubernetes
[kubernetes/website]: https://git.k8s.io/website

## Summary

This is to add additional access control over the existing impersonation action. An impersonator
who impersonates another user is required to have the additional permissions to impersonate on
certain group resources and verbs.

## Motivation
Today an impersonator can impersonate another user if the impersonator has the permission of
```yaml
- verbs: ["impersonate"]
  resources: ["users"]
  resourceNames: ["someUser"]
```
However, current impersonation mechanism is the "unrestricted impersonation", the impersonator can
get more power through impersonation. And because of this, it is strongly suggested not
to allow the controller to impersonate. More detailed discussion is in this
[issue](https://github.com/kubernetes/kubernetes/issues/27152)

There are use cases that needs a controller to be able to impersonate:

> 1. A controller impersonates a node the controller is running on. This happens when 

Reviewer 5: > which has the side-effect that the controller's identity is more or less "lost", so admission controllers have a hard time understanding if a request with username=system:node:foo really is node foo or some agent acting with its permissions

This is intentional.  We could embed the original user-info or otherwise make it available in the authz/admission layer, but we are choosing not to.

per-node agents (CNI plugins for instance), want to read pods on a given node instead
of unrestricted pod access.

> 2. A controller runs as a deputy, receives request from the user, and impersonates

Reviewer 7: Yes, I was more thinking what ownership model of the user you had in mind. How does the controller (impersonator) know which user to impersonate when performing a request? A special label called `owner` on the object? A field on the CRD?

This proposal is to introduce additional permissions for impersonation, so that any 
impersonator can impersonate in a more restricted way.

### Goals
* Define required permission rules for the impersonator to impersonate 
  on certain resource/verb.
* Provide clear example on what permission the impersonator and user should have for the
  impersonator to impersonate user's certain verbs on certain resources.
* This is an opt-in feature and existing impersonation flows keep working as-is.
* Client side semantics of impersonation is unchanged, the existing impersonation headers
are re-used and no new headers are added.

### Non-Goals
* Adding impersonator's info to user.Info, so authorization webhooks can know both info
of impersonated user and the requester.

## Proposal

Introduce a set of verbs with prefix of `impersonate-on:`, e.g. `impersonate-on:create` and
`impersonate-on:get`. The impersonator needs to have these verbs with certain resources to
impersonate. 

Introduce verbs `impersonate:user-info`, `impersonate:serviceaccount`， `impersonate:node` and
`impersonate:scheduled-node`:
- `impersonate:user-info` limits the impersonator to impersonate users with 

> certain names/groups/userextras. The resources must be `users`/`groups`/`userextras`/`uids`.

Reviewer 5: The concept of API groups did not exist when the original impersonation for `users` and `groups` was written...

> certain names/groups/userextras. The resources must be `users`/`groups`/`userextras`/`uids`.

Reviewer 7: Yeah, I know, it's unfortunate.

I was more getting at how we communicate it to the users. Do we send a warning if we see that someone tries to do `impersonate authentication.k8s.io users` and/or `impersonate:user-info "" users`, i.e. applying the new way to the old verb, or old way to the new verb?

Or do we strive to unify this by actually sending a SAR with the new, consistent format (`impersonate authentication.k8s.io users`) by default, and then falling back to the legacy (`impersonate "" users`) for backwards compability, such that we can make documentation uniform?

```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    resource: users
    name: someUser
    verb: impersonate
  user: impersonator
```

The impersonator does not need the permission for the target action.

#### workflow when the feature is enabled

```mermaid
flowchart TD
  A[receive request] -->|feature enabled| B{Identify the target user via impersonation header}
  B --> |Impersonate-User has the prefix of system:node|C{The requester is on the same node}
  C --> |yes| D{Requester is authorized with impersonate:scheduled-node}
  C --> |no| E{Requester is authorized with impersonate:node}
  D --> |yes| F{Requester is authorized to impersonate the target action}
  D --> |no| E
  E --> |yes| F
  E --> |no| H
  F -->|yes| G[Allow]
  A -->|feature disable| H{Requester has the legacy impersonate permission}
  H -->|yes| G
  F -->|no| H
  B --> |Impersonate-User has the prefix of system:serviceaccount| I{Requester is authorized with impersonate:serviceaccount}
  I --> |yes| F
  I --> |no| H
  B --> J{Requester is authorized with impersonate:user-info}
  J --> |yes|F
  J --> |no|H
```

### User Stories (Optional)

#### Story 1

> As a controller, I want to impersonate node I am running on to list/get pods on the node. The service account

Reviewer 5: You can directly set the impersonation headers in the kubeconfig, and a daemonset running all or some of the operations as the node already has to have logic to read the node's kubeconfig.  So this is effectively a configuration change.

  - ""
  resources:
  - pods
  verbs:
  - impersonate-on:list
  - impersonate-on:get
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: impersonate:node:list
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: impersonate:node:list
subjects:
- kind: ServiceAccount
  name: default
  namespace: default
```

#### Story 2

As a controller, I am working as a deputy, receiving any user's request to access virtual machine console
in the `default` namespace.

```yaml
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRole
metadata:
  name: impersonate:vm:console
rules:
- apiGroups:
  - authentications.k8s.io
  resources:
  - users
  verbs:
  - impersonate:user-info
---
apiVersion: rbac.authorization.k8s.io/v1
kind: ClusterRoleBinding
metadata:
  name: impersonate:vm:console
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: impersonate:vm:console
subjects:
- kind: ServiceAccount
  name: deputy
  namespace: deputy-ns
---
apiVersion: rbac.authorization.k8s.io/v1
kind: Role
metadata:
  name: impersonate:vm:console
  namespace: default
rules:
  - apiGroups:
    - subresources.kubevirt.io
    resources:

>     - virtualmachines/console

Reviewer 5: > that could be authorized with a ValidatingAdmissionPolicy

Not sure I follow how this would work.

---

The intention is to run the full authz/admission chain as the impersonated user instead of granting any of those permissions directly to the deputy.

    resource: users
    name: someUser
    verb: impersonate:user-info
  user: impersonator
```
will be sent to the authorizer

#### Header `Impersonate-Group` is set

A subjectaccessreview
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentications.k8s.io
    resource: groups
    name: someGroup
    verb: impersonate:user-info
  user: impersonator
```
will be sent to the authorizer

#### Header `Impersonate-Uid` is set

A subjectaccessreview
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentications.k8s.io
    resource: uids
    name: someUID
    verb: impersonate:user-info
  user: impersonator
```
will be sent to the authorizer

#### Header with prefix `Impersonate-Extra-` is set

A subjectaccessreview
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentications.k8s.io
    resource: userextras
    subresource: extraKey
    name: extraValue
    verb: impersonate:user-info
  user: impersonator
```
will be sent to the authorizer

### Verb `impersonate:serviceaccount`

> Same as legacy impersonation, when the request has the header of `Impersonate-User`, and the value of the header

Reviewer 5: We would continue to do that, and likely do the same for `system:nodes` for impersonated nodes.

> Same as legacy impersonation, when the request has the header of `Impersonate-User`, and the value of the header

Reviewer 4: Clarify if the impersonator required to fully-specify the serviceaccount+groups(+uid?), or forbidden from specifying any impersonation headers other than the Impersonate-User: "system:serviceaccounts:$namespace:$name", and if the latter, clarify that the service account groups and uid (which I guess requires looking up the service account?) are auto-added into the impersonated user info

are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

Yes, this will be covered in the unit tests and integration tests.

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

There is not impact on rollout, the impersonator with existing impersonate permission can still perform the action.
When the system rollback, impersonator with `impersonate-on:` and `impersonate:` permission will no longer
be authorized to impersonate. Impersonator will need to have the unscoped impersonate permission.

###### What specific metrics should inform a rollback?

<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->
authorization_attempts_total shows greatly increased number.

> authorization_duration_seconds_bucket shows greatly increased number of request

Reviewer 5: These tell you about the `WithAuthorization` request handler, not actual authz calls.  Those are only handled by the `InstrumentedAuthorizer` which only has `apiserver_authorization_decisions_total` today.  We will need to enhance that to include `apiserver_authorization_duration_seconds`.  _Maybe_ we would add labels to distinguish between impersonation calls based on the verb?

This proposal is to introduce additional permissions for impersonation, so that any 
impersonator can impersonate in a more restricted way.

### Goals
* Define required permission rules for the impersonator to impersonate 
  on certain resource/verb.
* Provide clear example on what permission the impersonator and user should have for the
  impersonator to impersonate user's certain verbs on certain resources.
* This is an opt-in feature and existing impersonation flows keep working as-is.
* Client side semantics of impersonation is unchanged, the existing impersonation headers
are re-used and no new headers are added.

### Non-Goals
* Adding impersonator's info to user.Info, so authorization webhooks can know both info
of impersonated user and the requester.

## Proposal

Introduce a set of verbs with prefix of `impersonate-on:`, e.g. `impersonate-on:create` and
`impersonate-on:get`. The impersonator needs to have these verbs with certain resources to
impersonate. 

Introduce verbs `impersonate:user-info`, `impersonate:serviceaccount`， `impersonate:node` and
`impersonate:scheduled-node`:
- `impersonate:user-info` limits the impersonator to impersonate users with 

> certain names/groups/userextras. The resources must be `users`/`groups`/`userextras`/`uids`.

Reviewer 4: in the flowchart of authz checks that will be done, make it clear what happens if a request that impersonates multiple attributes arrives.

For example:
```
GET /api/v1/namespaces/ns/pods
Headers:
  Impersonate-User: a
  Impersonate-Uid: b
  Impersonate-Group: c
  Impersonate-Group: d
  Impersonate-Extra-foo: e
  Impersonate-Extra-foo: f
  Impersonate-Extra-bar: g
  Impersonate-Extra-bar: h
```

For that request, I think would result in these checks:
* constrained: impersonate:user-info (8 authz checks) && impersonate-on:list pods in namespace `ns` (1 authz check)
* unconstrained: impersonate (8 authz checks)

Questions I wasn't sure about:
* What order are constrained / unconstrained checks done in?
* For constrained, do we do the 8 impersonate:user-info checks or the single impersonate-on check first?
* If any of the 8 impersonate:user-info or impersonate checks fail, do we short-circuit the others (I hope we do)

Reviewer 5: > the single impersonate-on check first

Seems like we should do this part first for each of the new cases.

Author: > If any of the 8 impersonate:user-info or impersonate checks fail, do we short-circuit the others (I hope we do)
yes if any fails, it will fallback to legacy mode. There is a special case for `impersonate:scheulde-node`
1. check the `impersonate:scheulde-node`, fall back to `impersonate:node`
2. check `impersonate:node`, fallback to `impersonate`

> If any of the 8 impersonate:user-info or impersonate checks fail, do we short-circuit the others (I hope we do)
yes

This is the worst case:
1. enable the featuregate
3. the legacy impersonate permission is set with UserExtra as the resource.
4. all impersonate-as header is set
And with such, comparing with legacy mode, it will introduce N+1 checks.

> What order are constrained / unconstrained checks done in?
The plan is do constrained at first than unconstrained. The reason is that if user enables this feature, they should understand that they need to use the constrained permission. And if they does that, unconstrained checks will fail for every request even if it passes constrained checks.

Reviewer 4: > The reason is that if user enables this feature, they should understand that they need to use the constrained permission. And if they does that, unconstrained checks will fail for every request even if it passes constrained checks.

That doesn't sound correct / compatible… can you describe what you mean by that? The feature will be enabled by default as it graduates to stable, and we won't break the ability to grant unconstrained impersonation

Author: No, I don't mean we will break the ability to grant unconstrained impersonation. 

I think we should check the constrained impersonation first and if fails, fallback to unconstrained impersonation. And the reason is based on the assumption that user will likely to use constrained permission with this feature enabled and remove the legacy unconstrained permission. In such case, checking constrained permission would be appropriate. 

If the assumption is mostly likely user enables this feature but still use legacy permission, checking unconstrained permission would be more appropriate.

and the user must not be a node (username with a prefix of `system:node:`) and the user must
not be a service account (username with a prefix of `system:serviceaccount:`)
The resource names must be usernames, group names or values in the user extras accoringly.
- `impersonate:serviceaccount` that limits the impersonator to impersonate the serviceaccount with
the certain name/namespace. The resource must be `serviceaccounts`.
- `impersonate:node` that limits the impersonator to impersonate the node only. The resource
must be `nodes`, and the resourceName should be the name of the node. The impersonator must have this
verb to impersonate a node.
- `impersonate:scheduled-node` that limits the impersonator to impersonate the node the 
impersonator is running on. The resources must be `nodes`. For a controller impersonating the node
that it is running on, it will need to know the node name obtained via downward API:
```yaml
env:
- name: MY_NODE_NAME
  valueFrom:
    fieldRef:
    fieldPath: spec.nodeName
```
and then set in the kubeconfig:
```go
kubeConfig, _ := clientcmd.BuildConfigFromFlags("", "")
kubeConfig.Impersonate = rest.ImpersonationConfig{

> 	UserName: "system:node:" + os.Getenv("MY_NODE_NAME"),

Reviewer 4: Would they also have to add `Groups: ["system:authenticated","system:nodes"]` or not?

permission is also required for impersonator, the component will not get more
power when permission of `impersonate-on:` is given.

#### High request volume leads to high load on authorization chain.

For the scheduled node case, it is possible that a high request volume node
agent could be constantly performing impersonated requests, each of which would
add two extra authorization checks. This could put load on the authorization
chain. Comparing to other alternatives which would also introduce at least
one authorization check for each request call, the proposal is able to
satisfy more use case.

#### Delegating permission of impersonating wildcard action or wildcard subjects is not supported

Permissions must be delegated using specified verbs with the prefix of `impersonate:` and
`impersonate-on:`. Expressing a permission to impersonate ANY subject or ANY action is not
possible, since the authorization model does not have the concept of partial wildcard on verbs.
It should be ok since all these permission for constrained impersonation are strictly additive which
means it should be added when certain impersonation requests are needed.

> ## Design Details

Reviewer 4: Because `impersonate-on:...` permissions are granted independent of `impersonate:user-info` / `impersonate:serviceaccount` / `impersonate:node`, an impersonator cannot be granted independent `impersonate:... + impersonate-on:...` combinations.

Instead, their abilities expand to the union of all `impersonate:...` identities, constrained by union of all `impersonate-on:...` permissions.

@deads2k @enj is that well understood / considered acceptable? I didn't see any discussion of that in the design.

Reviewer 4: practically, this means if you granted something like:

```
* impersonate:scheduled-node + impersonate-on:write pods/status
* impersonate:serviceaccount:myns:my-ingress-controller + impersonate-on:get secrets
```

what that would actually grant is read access to every secret the scheduled-node had access to

Reviewer 5: I don't think I thought about the union case.  Seems like it would be better to further scope the `-on` check such that you would need to do:

```
Node pod status update stuff
* impersonate:scheduled-node
* impersonate-on:scheduled-node:update "pods/status"

Controller secret read stuff
* impersonate:serviceaccount "myns/my-ingress-controller"
* impersonate-on:serviceaccount:get "secrets"
```

@deads2k WDYT?

Reviewer 5: Note that it is still not a true intersection, as this causes a union still:

```
Controller pod read stuff
* impersonate:serviceaccount "myns/my-pod-controller"
* impersonate-on:serviceaccount:get "pods"

Controller secret read stuff
* impersonate:serviceaccount "myns/my-ingress-controller"
* impersonate-on:serviceaccount:get "secrets"
```

Maybe we move the name into verb itself like `impersonate[-on]:serviceaccount:<namespace/name>`?

```
Controller pod read stuff
* impersonate:serviceaccount:myns/my-pod-controller
* impersonate-on:serviceaccount:myns/my-pod-controller:get "pods"

Controller secret read stuff
* impersonate:serviceaccount:myns/my-ingress-controller
* impersonate-on:serviceaccount:myns/my-ingress-controller:get "secrets"
```

The implementation is no more difficult, but the RBAC is kind of ugly to write.

  user: impersonator
```
will be sent to the authorizer for each key and value pair.

### Verb `impersonate:serviceaccount`

Same as legacy impersonation, when the request has the header of `Impersonate-User`, and the value of the header
has a prefix of `system:serviceaccount:`, apiserver will check verb `impersonate:serviceaccount` with the authorizer
using the subjectaccessreview:
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentication.k8s.io
    resource: serviceaccounts
    name: serviceaccount-name
    namespace: serviceaccount-namespace
    verb: impersonate:serviceaccount
  user: impersonator
```

### Verb `impersonate:node` and `impersonate:scheduled-node`

When the request has the header of `Impersonate-User`, and the value has the prefix of `system:nodes`,
this verb is checked instead of `impersonate:user-info`. The subjectaccessreview below will be sent
to the authorizer:
```yaml
apiVersion: authorization.k8s.io/v1
kind: SubjectAccessReview
spec:
  resourceAttributes:
    group: authentication.k8s.io
    resource: nodes
    name: someNode
    verb: impersonate:nodes
  user: impersonator

> ```

Reviewer 4: Is the impersonator required to fully-specify the node+groups(+uid?), or forbidden from specifying any impersonation headers other than the Impersonate-User: "system:nodes:$name"? If the latter, are the node groups auto-added into the impersonated user info?

Reviewer 5: https://github.com/kubernetes/enhancements/pull/5285#discussion_r2157544911

`impersonate:scheduled-node` is a special verb to check when two conditions are met:
1. The impersonator is impersonating a node by setting the header `Impersonate-User` and the value has a prefix

> of `system:node:` on the request.

Reviewer 4: Is the impersonator required to fully-specify the node+groups(+uid?), or forbidden from specifying any impersonation headers other than the Impersonate-User: "system:nodes:$name"? If the latter, are the node groups auto-added into the impersonated user info?

> 2. The user info of the impersonator has an extra with key `authentication.kubernetes.io/node-name`, and the value

Reviewer 4: Is the impersonator required to be a service account, or just have the matching `authentication.kubernetes.io/node-name` extra info value?

- [Maturity levels (`alpha`, `beta`, `stable`)][maturity-levels]
- [Feature gate][feature gate] lifecycle
- [Deprecation policy][deprecation-policy]

Clearly define what graduation means by either linking to the [API doc
definition](https://kubernetes.io/docs/concepts/overview/kubernetes-api/#api-versioning)
or by redefining what graduation means.

In general we try to use the same stages (alpha, beta, GA), regardless of how the
functionality is accessed.

[feature gate]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[maturity-levels]: https://git.k8s.io/community/contributors/devel/sig-architecture/api_changes.md#alpha-beta-and-stable-versions
[deprecation-policy]: https://kubernetes.io/docs/reference/using-api/deprecation-policy/

Below are some examples to consider, in addition to the aforementioned [maturity levels][maturity-levels].
-->
#### Alpha

- Feature implemented behind a feature flag
- Initial e2e tests completed and enabled

#### Beta

- Determine if additional tests are necessary
- Ensure reliability of existing tests
- Determine if some mechanism should be introduced to reduce the extra permission checks

> 

Author: if you mean other authorizer like webhook authorizer, it is in the alpha scope as the subjectaccessreview will be sent to the authorizer.

One example is

```yaml
{
  "apiVersion": "authorization.k8s.io/v1beta1",
  "kind": "SubjectAccessReview",
  "spec": {
    "resourceAttributes": {
      "namespace": "default",
      "verb": "get",
      "group": "example.org",
      "resource": "something"
    },
    "accessRule": {
      "kind": "ClusterAccessRule",
      "group version": "example.org/v1",
      "name": "elevation-of-privilege"
    },
    "user": "impersonator",
    "group": []
    ]
  }
}
```
And authorizer checks the accessRule on whether a certain impersonate action is allowed. This is a more
complicated approach that requires changes on existing RBAC/SAR, while the current proposal does not
introduce change on RBAC/SAR.

#### Conditional Authorization

Conditional authorization is the emerging work to provide more complicated authorization policy
with CEL expressions. Potentially it would be able to reduce the number of permission checks for the impersonation
in this proposal. The work is still in very early stage, and will bring many changes in the exising authorization
model. It is possible to enhance constrained impersonation in this proposal with conditional authorization in the

> future.

Reviewer 7: I strive to start some of the more "formal" KEP design work around this in July, after my vacation. What I have now is https://speakerdeck.com/luxas/conditional-authorization-for-kubernetes-sig-auth-presentation, which I presented during the June 4 SIG Auth meeting. When the recording is up, you can check the discussion out.

I think it's key to balance expressiveness with analyzability (like [this](https://aws.amazon.com/blogs/opensource/introducing-cedar-analysis-open-source-tools-for-verifying-authorization-policies/)), such that we can build good introspecting / debugging / semantic analysing tools around this, to aid people in doing the right thing. E.g. we could issue warnings if the user does something potentially dangerous, and based on admin policy either block, or say "are you sure, this might be a bad idea?"
